src/idpi/operators/stencils.py

```diff
@@ -0,0 +1,110 @@
+"""Finite difference stencils on xarray dataarrays."""
+# Standard library
+from typing import Literal
+from typing import Protocol
+
+# Third-party
+import xarray as xr
+
+Dim = Literal["x", "y", "z"]
+
+STENCILS = {
+    "full": {
+        -1: slice(-2),
+        0: slice(1, -1),
+        1: slice(2, None),
+    },
+    "half": {
+        -1: slice(-1),
+        0: slice(None),
+        1: slice(1, None),
+    },
+}
+
+
+class Field(Protocol):
+    def dx(self) -> xr.DataArray:
+        ...
+
+    def dy(self) -> xr.DataArray:
+        ...
+
+    def dz(self) -> xr.DataArray:
+        ...
+
+
+class PaddedField:
+    def __init__(self, field: xr.DataArray):
+        self.indexes = field.indexes
+        self.field = field.pad({dim: 1 for dim in list("xyz")}, mode="edge")
+
+    def __getitem__(self, indices: tuple[int, int, int]) -> xr.DataArray:
+        i, j, k = indices
+        s = STENCILS["full"]
+        return self.field[{"x": s[i], "y": s[j], "z": s[k]}].assign_coords(self.indexes)
+
+    def dx(self) -> xr.DataArray:
+        return 0.5 * (self[1, 0, 0] - self[-1, 0, 0])
+
+    def dy(self) -> xr.DataArray:
+        return 0.5 * (self[0, 1, 0] - self[0, -1, 0])
+
+    def dz(self) -> xr.DataArray:
+        return 0.5 * (self[0, 0, 1] - self[0, 0, -1])
+
+
+class StaggeredField:
+    """Should have one more element along the staggered dimension."""
+
+    def __init__(self, field: xr.DataArray, padded: PaddedField, dim: Dim):
+        if field.indexes:
+            field = field.drop_indexes(field.indexes.keys())
+        self.dim = dim
+        self.field = field
+        self.padded = padded
+
+    def __getitem__(self, indices: tuple[int, int, int]) -> xr.DataArray:
+        i, j, k = indices
+        s = STENCILS["half"]
+        return self.field[{"x": s[i], "y": s[j], "z": s[k]}]
+
+    def dx(self):
+        if self.dim == "x":
+            return self[1, 0, 0] - self[-1, 0, 0]
```

Comment:
as mentioned previously I think operators of a framework should be composable and adhere to common, well described data model and syntax. Until now we were using the xarray isel by label indexing. In this PR, we combine the use of the xarray isel, a more numba style, and later in the tests bare numpy indexing. Letting this grow in a framework will lead to mix and unclear syntaxes and patterns for data access, with very different styles in various parts of the framework. I would suggest to adhere to the xarray isel until we decide to adopt another one, to avoid organic, nonuniform growth of styles.

Reply:
Can you provide some additional detail on how to achieve this?